### PR TITLE
HHH-17269 - Add hibernate.boot.allow_jdbc_metadata_access

### PR DESCRIPTION
**Pull request by sebersole**
https://hibernate.atlassian.net/browse/HHH-17269

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/cfg/JdbcSettings.java [hunk 1]
```diff
@@ -471,6 +471,22 @@ public interface JdbcSettings extends C3p0Settings, ProxoolSettings {
 	 */
 	String CONNECTION_HANDLING = "hibernate.connection.handling_mode";
 
+	/**
+	 * Whether access to JDBC {@linkplain java.sql.DatabaseMetaData metadata} during bootstrap is allowed.
+	 * <p/>
+	 * Typically, Hibernate accesses this metadata to understand the capabilities of the underlying
+	 * database to help minimize needed configuration.  Disabling this access means that only explicit
+	 * settings are used.  At a minimum, the Dialect to use must be specified using either the {@value #DIALECT}
+	 * or {@value JdbcSettings#JAKARTA_HBM2DDL_DB_NAME} setting.
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/cfg/JdbcSettings.java:480]: I think we should probably start trying to guess the dialect from the JDBC URL, at least for the major platforms we support. That's what we do in Reactive, IIRC.

Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/cfg/JdbcSettings.java:480]: Another thing: I would tell them to not only specify the dialect, but also the database version.

**Review by gavinking — APPROVED** [review]

File: hibernate-core/src/main/java/org/hibernate/cfg/JdbcSettings.java (hunk 1 above)
Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/cfg/JdbcSettings.java:480]: > I think we should probably start trying to guess the dialect from the JDBC URL

Wait, I think we actually _do_ do this but I guess only when using `DriverManager` directly.

Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/cfg/JdbcSettings.java:480]: > Another thing: I would tell them to not only specify the dialect, but also the database version.

That's reasonable to notate in the Javadoc.

> Wait, I think we actually do do this but I guess only when using DriverManager directly.

Right, when the URL would be specified anyway.  I think ex[ecting users who are using data sources to specify URL just for the purpose of resolving the Dialect to use is silly.  If they are (1) using a datasource and (2) want to disable metadata access on bootstrap - specifying the Dialect to use via either `DIALECT` or `JAKARTA_HBM2DDL_DB_NAME` is more than reasonable.

Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/cfg/JdbcSettings.java:480]: > That's reasonable to notate in the Javadoc.

Yup, that's what I mean.

>  I think expecting users who are using data sources to specify URL just for the purpose of resolving the Dialect to use is silly. 

Yeah sure, agreed.

**Comment by sebersole** [comment]
Squashed, rebased and integrated upstream.

**Pull request closed without merging**